File: examples/node/memcore-client.js
/**
 * Example memcore HTTP client (not a published SDK).
 * Uses Node 18+ global fetch. Never logs the API key.
 */

export class MemcoreApiError extends Error {
  constructor(status, payload) {
    const err = payload && typeof payload === "object" ? payload.error || {} : {};
    const code = err.code ? ` code=${err.code}` : "";
    const requestId = err.request_id ? ` request_id=${err.request_id}` : "";
    const message = err.message ? `: ${err.message}` : "";
    super(`memcore HTTP ${status}${code}${requestId}${message}`);
    this.name = "MemcoreApiError";
    this.status = status;
    this.payload = payload;
  }
}

export class MemcoreClient {
  constructor({ baseUrl, apiKey, orgId, timeoutMs = 30_000 }) {
    if (!baseUrl || !apiKey || !orgId) {
      throw new Error("baseUrl, apiKey, and orgId are required");
    }
    this.baseUrl = String(baseUrl).replace(/\/$/, "");
    this._apiKey = apiKey;
    this.orgId = orgId;
    this.timeoutMs = timeoutMs;
  }

  _headers() {
    return {
      Authorization: `Bearer ${this._apiKey}`,
      "X-Organization-ID": this.orgId,
      "Content-Type": "application/json",
      Accept: "application/json",
    };
  }

  async _request(method, path, { query, body } = {}) {
    const url = new URL(`${this.baseUrl}${path}`);
    if (query) {
      for (const [key, value] of Object.entries(query)) {
        if (value !== undefined && value !== null && value !== "") {
          url.searchParams.set(key, String(value));
        }
      }
    }

    const response = await fetch(url, {
      method,
      headers: this._headers(),
      body: body === undefined ? undefined : JSON.stringify(body),
      signal: AbortSignal.timeout(this.timeoutMs),
    });

    let payload;
    try {
      payload = await response.json();
    } catch {
      payload = { raw: "(non-JSON body)" };
    }

    if (!response.ok) {
      throw new MemcoreApiError(response.status, payload);
    }
    return payload;
  }

  createMemory({ userId, text }) {
    return this._request("POST", "/api/v1/memories", {
      body: {
        user_id: userId,
        messages: [{ role: "user", content: text }],
        metadata: { source: "examples/node" },
      },
    });
  }

  searchMemories({ userId, query, limit = 5 }) {
    return this._request("POST", "/api/v1/memories/search", {
      body: { user_id: userId, query, limit },
    });
  }

  buildContext({ userId, query, maxTokens = 1000 }) {
    return this._request("POST", "/api/v1/context", {
      body: { user_id: userId, query, max_tokens: maxTokens },
    });
  }

  listMemories({ userId, limit = 50, cursor }) {
    return this._request("GET", `/api/v1/users/${encodeURIComponent(userId)}/memories`, {
      query: { limit, cursor },
    });
  }

  deleteMemory({ userId, memoryId }) {
    return this._request(
      "DELETE",
      `/api/v1/users/${encodeURIComponent(userId)}/memories/${encodeURIComponent(memoryId)}`,
    );
  }

  exportUser({ userId }) {
    return this._request("GET", `/api/v1/users/${encodeURIComponent(userId)}/export`, {
      query: { include_events: "true" },
    });
  }

  importUser({ userId, payload, dryRun = false }) {
    return this._request("POST", `/api/v1/users/${encodeURIComponent(userId)}/import`, {
      body: { ...payload, dry_run: dryRun },
    });
  }
}

File: examples/python/memcore_client.py
"""Example memcore HTTP client (not a published SDK)."""

from __future__ import annotations

from typing import Any, Optional

import requests


class MemcoreApiError(Exception):
    def __init__(self, status_code: int, payload: Any) -> None:
        self.status_code = status_code
        self.payload = payload
        code = None
        message = None
        request_id = None
        if isinstance(payload, dict):
            err = payload.get("error") or {}
            if isinstance(err, dict):
                code = err.get("code")
                message = err.get("message")
                request_id = err.get("request_id")
        super().__init__(
            f"memcore HTTP {status_code}"
            + (f" code={code}" if code else "")
            + (f" request_id={request_id}" if request_id else "")
            + (f": {message}" if message else "")
        )


class MemcoreClient:
    def __init__(
        self,
        base_url: str,
        api_key: str,
        org_id: str,
        *,
        timeout: float = 30.0,
        session: Optional[requests.Session] = None,
    ) -> None:
        if not base_url or not api_key or not org_id:
            raise ValueError("base_url, api_key, and org_id are required")
        self.base_url = base_url.rstrip("/")
        self._api_key = api_key
        self.org_id = org_id
        self.timeout = timeout
        self._session = session or requests.Session()

    def _headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self._api_key}",
            "X-Organization-ID": self.org_id,
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

    def _request(self, method: str, path: str, **kwargs: Any) -> dict[str, Any]:
        url = f"{self.base_url}{path}"
        response = self._session.request(
            method,
            url,
            headers=self._headers(),
            timeout=self.timeout,
            **kwargs,
        )
        try:
            payload: Any = response.json()
        except ValueError:
            payload = {"raw": response.text[:500]}
        if not response.ok:
            raise MemcoreApiError(response.status_code, payload)
        if not isinstance(payload, dict):
            raise MemcoreApiError(response.status_code, {"error": {"message": "expected JSON object"}})
        return payload

    def create_memory(self, user_id: str, text: str) -> dict[str, Any]:
        return self._request(
            "POST",
            "/api/v1/memories",
            json={
                "user_id": user_id,
                "messages": [{"role": "user", "content": text}],
                "metadata": {"source": "examples/python"},
            },
        )

    def search_memories(self, user_id: str, query: str, limit: int = 5) -> dict[str, Any]:
        return self._request(
            "POST",
            "/api/v1/memories/search",
            json={"user_id": user_id, "query": query, "limit": limit},
        )

    def build_context(
        self, user_id: str, query: str, max_tokens: int = 1000
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            "/api/v1/context",
            json={"user_id": user_id, "query": query, "max_tokens": max_tokens},
        )

    def list_memories(self, user_id: str, *, limit: int = 50, cursor: Optional[str] = None) -> dict[str, Any]:
        params: dict[str, Any] = {"limit": limit}
        if cursor:
            params["cursor"] = cursor
        return self._request("GET", f"/api/v1/users/{user_id}/memories", params=params)

    def delete_memory(self, user_id: str, memory_id: str) -> dict[str, Any]:
        return self._request("DELETE", f"/api/v1/users/{user_id}/memories/{memory_id}")

    def export_user(self, user_id: str) -> dict[str, Any]:
        return self._request(
            "GET",
            f"/api/v1/users/{user_id}/export",
            params={"include_events": "true"},
        )

    def import_user(
        self, user_id: str, payload: dict[str, Any], dry_run: bool = False
    ) -> dict[str, Any]:
        body = dict(payload)
        body["dry_run"] = dry_run
        return self._request("POST", f"/api/v1/users/{user_id}/import", json=body)
